### PR TITLE
Feature: rest trigger

### PR DESCRIPTION
**Pull request by alexrodfe**
## Description

- Added rest trigger

## Motivation and Context

Users will be able to use a predefined rest trigger

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other changes (ci configuration, documentation or any other kind of changes)

## Checklist:

- [ ] I have created tests for my code changes, and the tests are passing.
- [ ] I have executed the pre-commit hooks locally.
- [ ] My change requires a change to the documentation (create a new issue if the documentation has not been updated).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: rest-trigger/go.mod [hunk 1]
```diff
@@ -0,0 +1,58 @@
+module github.com/konstellation-io/kai-processes/rest-trigger
+
+go 1.20
+
+require (
+	github.com/gin-gonic/gin v1.9.1
+	github.com/google/uuid v1.3.1
+	github.com/konstellation-io/kai-sdk/go-sdk v1.1.0
```
Comment (kafkaphoenix) [inline rest-trigger/go.mod:8]: update sdk

File: rest-trigger/minikube_build.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+CGO_ENABLED=0 go build -o build/process main.go
+minikube image build -t konstellation/demo-grpc-trigger . -f develop.Dockerfile -p kai-local
```
Comment (kafkaphoenix) [inline rest-trigger/minikube_build.sh:4]: typo

**Review by kafkaphoenix — APPROVED** [review]

File: rest-trigger/main.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	"github.com/konstellation-io/kai-sdk/go-sdk/sdk"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+func main() {
+	runner.
+		NewRunner().
+		TriggerRunner().
+		WithInitializer(func(ctx sdk.KaiSDK) {
```
Comment (kafkaphoenix) [inline rest-trigger/main.go:27]: can you change ctx for kaiSDK here too please?

File: rest-trigger/main.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+			ctx.Logger.Info("Initializer")
+		}).
+		WithRunner(restServerRunner).
+		WithFinalizer(func(ctx sdk.KaiSDK) {
```
Comment (kafkaphoenix) [inline rest-trigger/main.go:31]: and here

**Review by kafkaphoenix — DISMISSED** [review]

**Review by kafkaphoenix — APPROVED** [review]

File: rest-trigger/app.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+centralized_configuration:
+  global:
+    bucket: global
+  product:
+    bucket: product
+  workflow:
+    bucket: workflow
+  process:
+    bucket: process
```
Comment (dfinteligenz) [inline rest-trigger/app.yaml:21]: Add a sample configuration for the process

File: rest-trigger/main.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+func main() {
+	runner.
+		NewRunner().
+		TriggerRunner().
+		WithInitializer(func(kaiSDK sdk.KaiSDK) {
+			kaiSDK.Logger.Info("Initializer")
+		}).
```
Comment (dfinteligenz) [inline rest-trigger/main.go:29]: If the initializer is not used, we can remove it from the builder.

File: rest-trigger/main.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+		WithRunner(restServerRunner).
+		WithFinalizer(func(kaiSDK sdk.KaiSDK) {
+			kaiSDK.Logger.Info("Finalizer")
+		}).
```
Comment (dfinteligenz) [inline rest-trigger/main.go:33]: If the finalizer is not used, we can remove it from the builder.

File: rest-trigger/main.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func restServerRunner(tr *trigger.Runner, kaiSDK sdk.KaiSDK) {
+	kaiSDK.Logger.Info("Starting http server", "port", 8080)
+
+	bgCtx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer stop()
+
+	r := gin.Default()
+	path := "/trigger"
```
Comment (dfinteligenz) [inline rest-trigger/main.go:44]: Add the path as an optional parameter to be set in the process configuration, and set the default value as `trigger`

File: rest-trigger/README.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+ "method": "POST",
+ "body": {"keyA": "valueA", "keyB": "valueB"},
+}
+```
+
+### Trigger's Input from other processes (JSON)
+
+| Key         | Type | Value                                                                      |
+|-------------|------|----------------------------------------------------------------------------|
+| status_code | str  | Status code for the workflow execution, used later in http server response |
```
Comment (dfinteligenz) [inline rest-trigger/README.md:36]: Replace status_code for method and modify the description on the value.

File: rest-trigger/README.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+the trigger need to marshal protobuf based JSON messages.
+
+Both output and input messages towards this process require the usage of specific keys in the JSON.
+
+### Trigger's Output (JSON)
+
+| Key       | Type    | Value                                                                                     |
+|-----------|---------|-------------------------------------------------------------------------------------------|
+| method    | str     | The REST method used                                                                      |
+| body      | []byte  | Optional. For POST and PUT methods the body is taken from the request's body              |
```
Comment (dfinteligenz) [inline rest-trigger/README.md:21]: Add an optional path parameter for setting a custom parameter. Use the `/trigger` by default

**Review by dfinteligenz — DISMISSED** [review]

File: rest-trigger/README.md (hunk 9 above)
Comment (alexrodfe) [reply rest-trigger/README.md:54]: This value is correct. This doc is describing the input from another process, the message received in the trigger from another process. So status code is a variable taken as response to formulate the rest response in the trigger.

**Review by kafkaphoenix — APPROVED** [review]

File: rest-trigger/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    --disabled-password \
+    --gecos "" \
+    --home "/nonexistent" \
+    --shell "/sbin/nologin" \
+    --no-create-home \
+    --uid "${UID}" \
+    "${USER}"
+
+WORKDIR /app
+COPY app.yaml app.yaml
```
Comment (kafkaphoenix) [inline rest-trigger/Dockerfile:31]: remove this line we dont want to copy app yaml in production

**Review by kafkaphoenix — APPROVED** [review]

**Pull request merged**